### PR TITLE
ascii_data initialization for mesh deformation

### PR DESCRIPTION
**Pull request by ricitron**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

I added 'ascii_data' as an option for initializing the new mesh deformation interface. Initial topography can now be defined using the mesh_deformation interface with ascii data. 

### Before your first pull request:

* [X ] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [ X] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [X ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: include/aspect/mesh_deformation/ascii_data.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef _aspect_mesh_deformation_model_ascii_data_h
+#define _aspect_mesh_deformation_model_ascii_data_h
```
Comment (gassmoeller) [inline include/aspect/mesh_deformation/ascii_data.h:23]: As a convention we use the directory/file structure for the include guards, so your code works, but please change to be consistent with the rest of the files.

```suggestion
#ifndef _aspect_mesh_deformation_ascii_data_h
#define _aspect_mesh_deformation_ascii_data_h
```

File: include/aspect/mesh_deformation/ascii_data.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+
+namespace aspect
+{
+  namespace MeshDeformation
+  {
+    using namespace dealii;
+
+    /**
+     * A class that implements topography determined
```
Comment (gassmoeller) [inline include/aspect/mesh_deformation/ascii_data.h:37]:
```suggestion
     * A class that implements initial topography determined
```

File: include/aspect/mesh_deformation/ascii_data.h [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        /**
+         * Initialization function. This function is called once at the
+         * beginning of the program. Sets the boundary id of the surface boundary.
+         */
+        void
+        initialize () override;
+
+        // avoid -Woverloaded-virtual:
+        using Utilities::AsciiDataBoundary<dim>::initialize;
+
```
Comment (gassmoeller) [inline include/aspect/mesh_deformation/ascii_data.h:60]: Please remove one empty line (our convention is to have one empty line between functions in header files, and 3 lines in source files).

File: source/mesh_deformation/ascii_data.cc [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+namespace aspect
+{
+  namespace MeshDeformation
+  {
+    template <int dim>
+    AsciiData<dim>::AsciiData ()
+      :
+      surface_boundary_id(1)
+    {}
+
```
Comment (gassmoeller) [inline source/mesh_deformation/ascii_data.cc:41]: See my comment above, please adjust to have 3 empty lines between all the functions in this file. This was inconsistent in the initial topography model you took this from, but this is our chance to correct it :-).

File: source/mesh_deformation/ascii_data.cc [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    template <int dim>
+    void
+    AsciiData<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Mesh deformation");
+      {
+        Utilities::AsciiDataBase<dim>::declare_parameters(prm,
+                                                          "$ASPECT_SOURCE_DIR/data/boundary-velocity/ascii-data/test/",
```
Comment (gassmoeller) [inline source/mesh_deformation/ascii_data.cc:91]: Can you use the same default path and filename as here: https://github.com/geodynamics/aspect/blob/c115676cabdeafbde8ad86ea9678cd65d2b6f2da/source/geometry_model/initial_topography_model/ascii_data.cc#L124? Or did you choose the boundary velocity data for some specific reason?

**Review by gassmoeller — CHANGES_REQUESTED** [review]
This is awesome! When I merged #3751 I thought about your work during the hackathon and was planning to send an email when I had some time. This PR looks nearly ready, except for my minor comments, and please add a test case. For that you can copy the test parameter file and output of `tests/box_initial_topography_ascii_data` to a new name (e.g. box_initial_mesh_deformation_ascii_data) and upload, then replace the output with the output from the reference tester (the procedure is described here: http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests, but let us know if something does not work).

File: source/mesh_deformation/ascii_data.cc (hunk 5 above)
Comment (ricitron) [reply source/mesh_deformation/ascii_data.cc:91]: The box_2d_top.0.txt file referenced by the initial_topography model/ascii_data.cc is missing. I set it to the box_3d file now instead.

File: include/aspect/mesh_deformation/ascii_data.h (hunk 3 above)
Comment (ricitron) [reply include/aspect/mesh_deformation/ascii_data.h:60]: done

File: source/mesh_deformation/ascii_data.cc (hunk 4 above)
Comment (ricitron) [reply source/mesh_deformation/ascii_data.cc:40]: done

**Comment by ricitron** [comment]
Thanks! I made the formatting changes and added the test. Will see if the test works...

**Comment by tjhei** [comment]
You will have to update the test results https://jenkins.tjhei.info/blue/organizations/jenkins/aspect/detail/PR-3864/7/tests (download the diff from "artifacts" and use ``git apply`` as described in the manual).

File: source/mesh_deformation/ascii_data.cc [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        topography_direction = -gravity / gravity.norm();
+
+      return topo * topography_direction;
+    }
+
+
+
+    template <int dim>
+    Tensor<1,dim-1>
+    AsciiData<dim>::vector_gradient(const Point<dim> &point) const
```
Comment (tjhei) [inline source/mesh_deformation/ascii_data.cc:80]: Why do you need this function? I would think everything works after deleting it.

Comment (ricitron) [reply source/mesh_deformation/ascii_data.cc]: It was copied from the geometry ascii data implementation. I deleted it now.

**Comment by ricitron** [comment]
OK, thanks, I have updated the test results.

File: source/mesh_deformation/ascii_data.cc (hunk 6 above)
Comment (gassmoeller) [reply source/mesh_deformation/ascii_data.cc]: It looks like you only deleted the implementation in the .cc file, but you still have the declaration in the .h file. Please remove that as well. Afterwards this looks good to go.

Comment (ricitron) [reply source/mesh_deformation/ascii_data.cc]: OK, I fixed that now.

**Review by gassmoeller — APPROVED** [review]
Thanks! This is ready.

**Pull request merged**